connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTaskTest.java

```diff
@@ -651,6 +652,40 @@ public void testTopicCreateSucceedsWhenCreateReturnsNewTopicFound() {
         verifyTopicCreation();
     }
 
+    @Test
+    public void testSendRecordsRetriableException() {
+        createWorkerTask();
+
+        SourceRecord record1 = new SourceRecord(PARTITION, OFFSET, TOPIC, 1, KEY_SCHEMA, KEY, RECORD_SCHEMA, RECORD);
+        SourceRecord record2 = new SourceRecord(PARTITION, OFFSET, TOPIC, 2, KEY_SCHEMA, KEY, RECORD_SCHEMA, RECORD);
+        SourceRecord record3 = new SourceRecord(PARTITION, OFFSET, TOPIC, 3, KEY_SCHEMA, KEY, RECORD_SCHEMA, RECORD);
+
+        expectConvertHeadersAndKeyValue(emptyHeaders(), TOPIC);
+        expectTaskGetTopic();
+
+        when(transformationChain.apply(eq(record1))).thenReturn(null);
+        when(transformationChain.apply(eq(record2))).thenReturn(null);
+        when(transformationChain.apply(eq(record3))).thenReturn(record3);
+
+        TopicPartitionInfo topicPartitionInfo = new TopicPartitionInfo(0, null, Collections.emptyList(), Collections.emptyList());
+        TopicDescription topicDesc = new TopicDescription(TOPIC, false, Collections.singletonList(topicPartitionInfo));
+        when(admin.describeTopics(TOPIC)).thenReturn(Collections.singletonMap(TOPIC, topicDesc));
+
+        when(producer.send(any(), any())).thenThrow(new RetriableException("Retriable exception")).thenReturn(null);
+
+        workerTask.toSend = Arrays.asList(record1, record2, record3);
+
+        // The producer throws a RetriableException the first time we try to send the third record
+        assertFalse(workerTask.sendRecords());
+
+        // The next attempt to send the third record should succeed
+        assertTrue(workerTask.sendRecords());
+
+        // Ensure that the first two records that were filtered out by the transformation chain
+        // aren't re-processed when we retry the call to sendRecords()
+        verify(transformationChain, times(4)).apply(any(SourceRecord.class));
```

Comment:
Do you think it might be worth it to be more explicit in the verifications here? I.e., verify that `record1` and `record2` were transformed only once, but `record3` was transformed twice?

Also, it's a little strange that we're re-transforming a record for each retry. If it's not too much we may want to refine that logic to only transform each record at most once.

Reply:
Yeah I agree. An explicit check for `record1` and `record2` being transformed once and `record3` twice would be good.
Regarding the duplicate re-transformations, I guess it should be achievable by storing a mapping between `SourceRecord` and the `ProducerRecord` ie pre-transformed and transformed. We can keep clearing the map as and when the entire batch of records are processed. In theory, that could have been another way of achieving what we are trying to achieve in this PR i.e if SMT outputs a record's transformation as null, we can update the mapping and we can skip it for subsequent retries. But the solution in this PR is cleaner.

Reply:
Transformations / SMTs are applied on `SourceRecord` instances and also return `SourceRecord`s. Transformed records are converted to `ProducerRecord`s by converters. Although that does bring up the point that not only are we re-transforming a record for each retry, we're also re-converting each time. I agree that this does seem a little strange. 

We currently rely on the global `toSend` list to maintain the source records to be retried between successive calls to `sendRecords` (separated by a backoff period). I guess we could decouple the transformation + conversion of a batch of records from the loop where we dispatch the records - i.e. maintain a new global mapping from pre-transform `SourceRecord`s to transformed + converted `ProducerRecord`s which is populated first for each polled batch and then dispatched in `sendRecords` (following the same retry logic that we currently use). Although I feel like this optimization refactor could be a follow-up / separate item, WDYT?

Reply:
Yeah that global map approach will work. Having said that, I agree we can take it up as a separate PR and not complicate this PR.

Reply:
Giving it a second thought though, this bug only manifests because we're currently transforming and converting records in the same method that is called on each attempt to send a batch of records; so it doesn't seem unreasonable to change that behavior in this PR itself.